File: src/main/java/com/vaadin/demo/component/numberfield/NumberFieldValidation.java
package com.vaadin.demo.component.numberfield;

import com.vaadin.flow.component.html.Div;
import com.vaadin.flow.component.textfield.IntegerField;
import com.vaadin.flow.component.textfield.IntegerField.IntegerFieldI18n;
import com.vaadin.flow.router.Route;
import com.vaadin.demo.DemoExporter; // hidden-source-line

@Route("number-field-validation")
public class NumberFieldValidation extends Div {

    public NumberFieldValidation() {
        // tag::snippet[]
        IntegerField integerField = new IntegerField();
        integerField.setLabel("Quantity");
        integerField.setHelperText("Max 10 items");
        integerField.setRequiredIndicatorVisible(true);
        integerField.setMin(1);
        integerField.setMax(10);
        integerField.setValue(2);
        integerField.setStepButtonsVisible(true);

        integerField.setI18n(new IntegerFieldI18n()
            .setRequiredErrorMessage("Field is required")
            .setBadInputErrorMessage("Invalid number format")
            .setMinErrorMessage("Quantity must be at least 1")
            .setMaxErrorMessage("Maximum 10 items available"));

        add(integerField);
        // end::snippet[]
    }

    public static class Exporter extends DemoExporter<NumberFieldValidation> { // hidden-source-line
    } // hidden-source-line
}

File: src/main/frontend/demo/component/numberfield/number-field-validation.ts
import 'Frontend/demo/init'; // hidden-source-line
import '@vaadin/integer-field';
import { html, LitElement } from 'lit';
import { customElement, state } from 'lit/decorators.js';
import type { IntegerField, IntegerFieldValidatedEvent } from '@vaadin/integer-field';
import { applyTheme } from 'Frontend/generated/theme';

@customElement('number-field-validation')
export class Example extends LitElement {
  protected override createRenderRoot() {
    const root = super.createRenderRoot();
    // Apply custom theme (only supported if your app uses one)
    applyTheme(root);
    return root;
  }

  @state()
  private errorMessage = '';

  protected override render() {
    return html`
      <!-- tag::snippet[] -->
      <vaadin-integer-field
        label="Quantity"
        helper-text="Max 10 items"
        required
        min="1"
        max="10"
        value="2"
        step-buttons-visible
        .errorMessage="${this.errorMessage}"
        @validated="${(event: IntegerFieldValidatedEvent) => {
          const field = event.target as IntegerField;
          const { validity } = field.inputElement as HTMLInputElement;
          if (validity.badInput) {
            this.errorMessage = 'Invalid number format';
          } else if (validity.valueMissing) {
            this.errorMessage = 'Field is required';
          } else if (validity.rangeUnderflow) {
            this.errorMessage = `Quantity must be at least ${field.min}`;
          } else if (validity.rangeOverflow) {
            this.errorMessage = `Maximum ${field.max} items available`;
          } else {
            this.errorMessage = '';
          }
        }}"
      ></vaadin-integer-field>
      <!-- end::snippet[] -->
    `;
  }
}
